### PR TITLE
Omit mut from Tokenizer

### PR DESCRIPTION
**Pull request by kampersanda**
This PR implemented a stateless version of `Tokenizer` for multithreading.

### CODE REVIEW

**Comment by kampersanda** [comment]
@vbkaisetsu 

Could you review the current API design found in the `cargo doc` rather than the many file changes included in this PR?
If there are detailed modification points, could you give me them with another PR?

File: vibrato/src/state.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+use crate::errors::Result;
+use crate::sentence::Sentence;
+use crate::token::{Token, TokenIter};
+use crate::tokenizer::lattice::{Lattice, Node};
+
+/// Maintainer of an input sentence and tokenized results.
+///
+/// It also holds the internal data structures used in tokenization,
+/// which can be reused to avoid unnecessary memory reallocation.
+pub struct State<'a> {
```
Comment (vbkaisetsu) [inline vibrato/src/state.rs:13]: Shouldn't this struct be named `Lattice`?
The name `State` is difficult to understand for those who do not know the internal implementation.

Comment (kampersanda) [reply vibrato/src/state.rs:13]: @vbkaisetsu How about `Worker`?

I think the role of this structure is to provide a routine for tokenization. For many light users, the internal data structure will not be interesting, so `Worker` is more friendly than the technical name `Lattice`.

e.g.,

```rust
let tokenizer = vibrato::Tokenizer::new(dict);
let mut worker = tokenizer.new_worker();
worker.reset_sentence("京都東京都");
worker.tokenize();
worker.token(0);
...
```

Comment (vbkaisetsu) [reply vibrato/src/state.rs:13]: ok

File: vibrato/src/token.rs [hunk 2]
```diff
@@ -1,144 +1,106 @@
...
-
-    #[inline(always)]
-    fn index(&self, i: usize) -> usize {
-        self.nodes.len() - i - 1
-    }
-}
-
-/// Token.
+/// Resultant token.
 pub struct Token<'a> {
```
[the remark this answers is not in the document]
Comment (kampersanda) [reply vibrato/src/token.rs:8]: https://github.com/daac-tools/vibrato/pull/21/commits/d9986a248488c303d3c024f99ad95dcc9d8afafb

File: vibrato/src/tokenizer.rs [hunk 3]
```diff
@@ -85,59 +71,45 @@ impl<'a> Tokenizer<'a> {
...
-        self.lattice.append_top_nodes(&mut self.tokens.nodes);
+    /// Creates a new state.
+    pub fn new_state(&self) -> State {
+        State::new(&self.dict)
+    }
 
-        Ok(&self.tokens)
+    /// Tokenizes the input sentence set in `state`,
+    /// returning the result through `state`.
+    pub fn tokenize(&self, state: &mut State) {
```
[the remark this answers is not in the document]
Comment (kampersanda) [reply vibrato/src/tokenizer.rs]: It's a significant point. Thanks.
https://github.com/daac-tools/vibrato/pull/21/commits/c69049e2cd381f14cd295ba64c61fd3e97b143c2

File: vibrato/src/state.rs (hunk 1 above)
Comment (kampersanda) [reply vibrato/src/state.rs]: https://github.com/daac-tools/vibrato/pull/21/commits/c69049e2cd381f14cd295ba64c61fd3e97b143c2

**Review by vbkaisetsu — APPROVED** [review]

**Comment by kampersanda** [comment]
@vbkaisetsu Could you approve it again because of the merge?

**Review by vbkaisetsu — APPROVED** [review]

**Pull request merged**